### PR TITLE
Keep DFlash anchor blocks inside document boundaries

### PR DESCRIPTION
**Pull request by khazic**
## Problem

`select_anchors` only excludes the last `block_size` positions of the whole packed sequence, not of each document. An anchor chosen near the end of one document produces block slots whose target logits and loss mask come from the next document, while the attention mask restricts those queries to the previous document's context: the draft is trained to predict another document's tokens from unrelated context, corrupting the loss and the position-accuracy metrics for every boundary block.

## Fix

`select_anchors` accepts the packed sequence's `document_ids` (the call site already has them) and excludes anchors whose block would cross a document boundary: documents are contiguous, so a block stays inside one document iff its first and last positions share a document id. Behavior without `document_ids` is unchanged.

Note on scope: this changes anchor-selection behavior for packed batches (three files counting the test), so per CONTRIBUTING this one may warrant a look at the intended semantics; the previous behavior trains across document boundaries, which did not look intentional.

## Tests

New unit tests assert no selected block spans two documents, the exact count of valid anchors for a two-document layout, and unchanged behavior when `document_ids` is omitted. The boundary test fails on main.

### CODE REVIEW

**Review by shanjiaz — COMMENTED** [review]
The fix is clean and correct, but just to explain that this probably doesn't actually affect correctness. First of all, anchors only land on loss_mask=1 positions (assistant response). The next document in a packed sequence always starts with prompt tokens (loss_mask=0), so cross-boundary positions in the block won't contribute to the loss, the model never learns wrong pattern across documents. `base_prefix_mod` in `attention.py` already enforces document boundaries for base-sequence KV attention. The within-block path `same_block_mod` attends to other draft tokens (mask embeddings), not to the base model's hidden states, so cross-boundary positions don't leak information from the neighboring document. Their loss is also masked out. If you agree, can we close this PR?

File: src/speculators/models/dflash/utils.py [hunk 1]
```diff
@@ -43,6 +48,14 @@ def select_anchors(
 
     valid_mask = loss_mask.bool().clone()
     valid_mask[:, -block_size:] = False
+    if document_ids is not None and block_size > 1:
+        # An anchor's block spans positions [p, p + block_size - 1]; documents
+        # are contiguous, so the block stays in one document iff its first and
+        # last positions share a document id.
+        same_doc = (
+            document_ids[:, : -(block_size - 1)] == document_ids[:, block_size - 1 :]
```
[the remark this answers is not in the document]
Comment (khazic) [reply src/speculators/models/dflash/utils.py]: Confirmed and fixed: the checked horizon is now block_size when sample_from_anchor is set (DSpark) and block_size - 1 otherwise, so it also applies at block_size == 1. Tests cover both.

**Review by speculatorsbot — COMMENTED** [review]
Clean, focused fix. The `horizon` calculation correctly distinguishes DFlash (`block_size - 1`) from DSpark (`block_size`), the contiguity-based boundary check (`document_ids[:, :-horizon] == document_ids[:, horizon:]`) avoids materializing per-anchor masks, and the four new tests cover the key cases including the `block_size == 1` edge case for DSpark.

**On @shanjiaz's argument**: The observation that cross-boundary positions are typically `loss_mask=0` (prompt tokens at the start of the next document) is correct for standard chat fine-tuning data. Under that convention, `aligned_loss_mask` zeros out cross-boundary slots, so they contribute to neither the loss nor the position-accuracy metrics (`compute_accuracy_multi_step` applies the same mask). The attention-level isolation via `base_prefix_mod` also holds independently of this fix.

However, `loss_mask=0` at every document boundary is a data-packing convention, not an invariant the training loop enforces. The collator constructs `document_ids` from lengths and `loss_mask` from role offsets independently — there is no check that a document's first token is always `loss_mask=0`. Pre-tokenized data (`_carry_pretokenized` in `preprocessing.py`) passes user-supplied masks through without validation. If someone trains on data where adjacent packed documents both begin with `loss_mask=1` tokens (pre-training corpora, or chat datasets with full-sequence loss), the cross-boundary slots would carry active loss on targets from an unrelated document. This fix removes that assumption at negligible cost (one element-wise comparison + bitwise AND on the valid mask), so it seems worth keeping as defensive correctness.

🤖 Generated with [Claude Code](https://claude.ai/code) using the `/pr-review` skill

**Comment by khazic** [comment]
@shanjiaz agreed, closing. Your reasoning holds for the data this trains on: anchors only land on `loss_mask=1` positions, the next packed document starts with prompt tokens, and `base_prefix_mod` already enforces document boundaries for base-sequence KV attention, so the cross-boundary slots neither contribute to the loss nor leak hidden states. Thanks for taking the time to spell it out.
